node/node.go

```diff
@@ -614,95 +614,83 @@ func (n *Node) waitRole(ctx context.Context, role string) error {
+		ForceNewCluster: n.config.ForceNewCluster,
+		RemoteAPI: manager.RemoteAddrs{
+			ListenAddr:    n.config.ListenRemoteAPI,
+			AdvertiseAddr: n.config.AdvertiseRemoteAPI,
+		},
+		ControlAPI:     n.config.ListenControlAPI,
+		SecurityConfig: securityConfig,
+		ExternalCAs:    n.config.ExternalCAs,
+		JoinRaft:       remoteAddr.Addr,
+		StateDir:       n.config.StateDir,
+		HeartbeatTick:  n.config.HeartbeatTick,
+		ElectionTick:   n.config.ElectionTick,
+	})
+	if err != nil {
+		return err
+	}
 
+	done := make(chan struct{})
+	var runErr error
+	go func() {
+		runErr = m.Run(context.Background())
+		close(done)
+	}()
+	defer func() {
+		n.setControlSocket(nil)
 		n.Lock()
-		n.manager = m
+		n.manager = nil
 		n.Unlock()
+		m.Stop(ctx)
+		<-done
+	}()
 
-		connCtx, connCancel := context.WithCancel(ctx)
-		go n.initManagerConnection(connCtx, ready)
-
-		// this happens only on initial start
-		if ready != nil {
-			go func(ready chan struct{}) {
-				select {
-				case <-ready:
-					addr, err := n.RemoteAPIAddr()
-					if err != nil {
-						log.G(ctx).WithError(err).Errorf("get remote api addr")
-					} else {
-						n.remotes.Observe(api.Peer{NodeID: n.NodeID(), Addr: addr}, remotes.DefaultObservationWeight)
-					}
-				case <-connCtx.Done():
-				}
-			}(ready)
-			ready = nil
-		}
+	n.Lock()
+	n.manager = m
+	n.Unlock()
 
-		roleChanged := make(chan error)
-		waitCtx, waitCancel := context.WithCancel(ctx)
-		go func() {
-			err := n.waitRole(waitCtx, ca.WorkerRole)
-			roleChanged <- err
-		}()
+	connCtx, connCancel := context.WithCancel(ctx)
+	defer connCancel()
 
-		select {
-		case <-done:
-			// Fail out if m.Run() returns error, otherwise wait for
-			// role change.
-			if runErr != nil {
-				err = runErr
-			} else {
-				err = <-roleChanged
+	go n.initManagerConnection(connCtx, ready)
+
+	// this happens only on initial start
+	if ready != nil {
+		go func(ready chan struct{}) {
+			select {
+			case <-ready:
+				addr, err := n.RemoteAPIAddr()
+				if err != nil {
+					log.G(ctx).WithError(err).Errorf("get remote api addr")
+				} else {
+					n.remotes.Observe(api.Peer{NodeID: n.NodeID(), Addr: addr}, remotes.DefaultObservationWeight)
+				}
+			case <-connCtx.Done():
 			}
-		case err = <-roleChanged:
-		}
+		}(ready)
+	}
 
-		n.Lock()
-		n.manager = nil
-		n.Unlock()
+	select {
+	case <-done:
+		return runErr
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
 
-		select {
-		case <-done:
-		case <-ctx.Done():
-			err = ctx.Err()
-			m.Stop(context.Background())
-			<-done
+func (n *Node) superviseManager(ctx context.Context, securityConfig *ca.SecurityConfig, ready chan struct{}) error {
+	for {
+		if err := n.waitRole(ctx, ca.ManagerRole); err != nil {
+			return err
 		}
-		connCancel()
-		n.setControlSocket(nil)
-		waitCancel()
-
-		if err != nil {
+		if err := n.runManager(ctx, securityConfig, ready); err != nil {
+			return err
+		}
+		ready = nil
+		if err := n.waitRole(ctx, ca.WorkerRole); err != nil {
```

Comment:
What if another promotion has happened before this is called. Wouldn't it block here then?